### PR TITLE
Add characters and words counts to editor

### LINKED ISSUE #1474: Show the number of characters in translation
As @tobifjellner suggests [here](https://wordpress.slack.com/archives/C02RP50LK/p1660119378388429), it would be interesting to add the number of characters in the translation, in the bottom right corner.

### PR DESCRIPTION
**Pull request by pedro-mendonca**
## What?
Add Characters and Words counting to string editor

Fixes #1474

## How?
Add counts below the string, on opening the editor row or when the input data changes, for every singular/plural textareas.
- [x] Add counts HTML on textarea focus to add counts on opening the editor row
- [x] Update counts HTML on textarea input change
- [x] All editor row textareas (singular and plurals) counts are updated on focus or input string change
- [x] Only the opened editor rows have the counts added/updated with JS, removing server process on page load.
- [x] The count uses WP core package [Wordcount](https://github.com/WordPress/gutenberg/tree/trunk/packages/wordcount) ([suggested](https://github.com/GlotPress/GlotPress/pull/1478#discussion_r958102966) by @ocean90)
- [x] The Wordcount package ignores HTML tags and supports word counts for Asian languages to count every character as words
- [x] The words count now depends on each `GP_Locale->word_count_type` setting added in #1482

## Screenshots or screencast 
imagem

~This JavaScript isn't prepared for i18n yet to avoid conflicts with previous PR #1369
Will be fixed after the PR merge.~
This JavaScript is already prepared for i18n.

### CODE REVIEW

File: assets/js/editor.js
Comment (ocean90) [inline assets/js/editor.js]:
```suggestion
					.on( 'input', 'tr.editor textarea.foreign-text', $gp.editor.hooks.update_count );
```

`input` should catch all these events.

File: gp-templates/helper-functions.php
Comment (ocean90) [inline gp-templates/helper-functions.php]: Do we have to replace newline characters first to avoid entries with only a single space? Looking at https://github.com/WordPress/wordpress-develop/blob/0d7487bad97e185c246e48be5a24fab5ee7bb985/src/wp-includes/formatting.php#L3950-L3951.

File: gp-templates/helper-functions.php
Comment (ocean90) [inline gp-templates/helper-functions.php]:
```suggestion
				number_format_i18n( $characters_count )
```

File: gp-templates/helper-functions.php
Comment (ocean90) [inline gp-templates/helper-functions.php]:
```suggestion
				number_format_i18n( $words_count )
```

File: gp-templates/helper-functions.php
Comment (ocean90) [reply gp-templates/helper-functions.php]: This also doesn't handle languages where word count is based on single characters.

File: assets/js/editor.js
Comment (ocean90) [inline assets/js/editor.js]: Should we make use of https://github.com/WordPress/gutenberg/tree/trunk/packages/wordcount?

**Review by ocean90 — CHANGES_REQUESTED** [review]

File: gp-templates/helper-functions.php
Comment (pedro-mendonca) [reply gp-templates/helper-functions.php]: > Do we have to replace newline characters first to avoid entries with only a single space?

Doesn't `\s` already match `\n`, `\r` and `\t`?

> This also doesn't handle languages where word count is based on single characters.

For that GlotPress needs to rely on the site language (core dev .po), should we go for it?

File: assets/js/editor.js
Comment (pedro-mendonca) [reply assets/js/editor.js]: It could be helpful, ~but it also means bumping JS to ES2015, adding the first JS dependency, and updating the deployment to include it.~ adding `wp-wordcount` handle as `gp-editor` dependency?

File: gp-templates/helper-functions.php
Comment (pedro-mendonca) [reply gp-templates/helper-functions.php]: @ocean90 what about completely remove the PHP fill of this data, and add it only on focus/update the current editor textarea with the `Wordcount` JS package?

**Comment by pedro-mendonca** [comment]
@ocean90 The PR already uses the Gutenberg/core package `Wordcount`.

Also handles languages where word count is based on single characters, but reads the setting from the site language settings.

It would make much more sense to use a GP_Locale property to allow reading wordcounts based on the translation set instead of one single site setting for all GP_Locales.
Example: For an asian translation set, it should count characters as words, but german the word count should be based on letters separated by spaces. Currently, if the site locale is set to `words`, it will read the wordcount as `words` for all Locales.

Would it make sense to add it as a GP_Locale property instead?

File: assets/js/editor.js
Comment (pedro-mendonca) [reply assets/js/editor.js]: Done, the PR is using Wordcount package.

File: gp-templates/helper-functions.php
Comment (pedro-mendonca) [reply gp-templates/helper-functions.php]: Removed the PHP logic and relying completely on the the Wordcount package to update the data on textarea focus or any input change. This reduces the PHP processing for all the strings and only processes the count on opening the row editor.

**Comment by ocean90** [comment]
> Also handles languages where word count is based on single characters, but reads the setting from the site language settings.

For the translation that definitely doesn't make sense so having this setting as part of `GP_Locale` sounds good to me. Maybe as a separate PR as done in #1479?

**Comment by pedro-mendonca** [comment]
> Maybe as a separate PR as done in https://github.com/GlotPress/GlotPress/pull/1479?

Yes, I'll wait for it to be merged. Will have to check wich alphabets use the words/characters for count.

**Comment by pedro-mendonca** [comment]
@ocean90 do you mean we should add a property `word_count_type` to `GP_Locale`? Or we could just rely on the alphabet for it?

**Comment by amieiro** [comment]
> > Maybe as a separate PR as done in #1479?
> 
> Yes, I'll wait for it to be merged. Will have to check wich alphabets use the words/characters for count.

Done @pedro-mendonca

**Comment by pedro-mendonca** [comment]
Also done, added the property in #1482

**Comment by pedro-mendonca** [comment]
@ocean90 the PR now reads the [newly added](https://github.com/GlotPress/GlotPress/pull/1482) `GP_Locale` `word_count_type` and every Locale editor behaves accordingly.
I think it's done.

**Review by ocean90 — APPROVED** [review]

**Pull request merged**
